kamu-adapter-http/src/http_server_simple_transfer_protocol.rs

```diff
@@ -0,0 +1,162 @@
+
+    axum::response::Response::builder()
+        .body(axum::body::boxed(body))
+        .unwrap()
+}
+
+pub async fn dataset_checkpoints_handler(
+    local_dataset_repository: axum::extract::Extension<Arc<dyn LocalDatasetRepository>>,
+    axum::extract::Path(params): axum::extract::Path<HashMap<String, String>>
+) -> axum::response::Response {
+
+    let dataset = resolve_dataset(local_dataset_repository, &params)
+        .await
+        .unwrap();
+
+    let physical_hash_param = params.get(PARAMETER_PHYSICAL_HASH).unwrap();
+    let physical_hash = Multihash::from_multibase_str(physical_hash_param.as_str()).unwrap();
+    let checkpoint_stream = dataset.as_checkpoint_repo().get_stream(&physical_hash).await.unwrap();
+    let body = axum_extra::body::AsyncReadBody::new(checkpoint_stream);
+
+    axum::response::Response::builder()
+        .body(axum::body::boxed(body))
+        .unwrap()
+}
+
+pub async fn dataset_push_ws_upgrade_handler(
+    ws: axum::extract::ws::WebSocketUpgrade,
+    local_dataset_repository: axum::extract::Extension<Arc<dyn LocalDatasetRepository>>,
+    axum::extract::Path(params): axum::extract::Path<HashMap<String, String>>
+) -> axum::response::Response {
+
+    let dataset = resolve_dataset(local_dataset_repository, &params)
+        .await
+        .unwrap();
+
+    ws.on_upgrade(
+        |socket| ws_smart_transfer_protocol_axum_server::dataset_push_ws_handler(socket, dataset)
+    )
+}
+
+pub async fn dataset_pull_ws_upgrade_handler(
+    ws: axum::extract::ws::WebSocketUpgrade,
+    local_dataset_repository: axum::extract::Extension<Arc<dyn LocalDatasetRepository>>,
+    axum::extract::Path(params): axum::extract::Path<HashMap<String, String>>,
+    TypedHeader(host): TypedHeader<Host>,
+) -> axum::response::Response  {
+
+    let dataset = resolve_dataset(local_dataset_repository, &params)
+        .await
+        .unwrap();
+
+    let host_ptr = Box::from(host.clone());
+    
+    ws.on_upgrade(
+        move |socket| {
+            let mut prefix_url_str = String::from("http://");
+            prefix_url_str += host_ptr.hostname();
+            if let Some(port) = host_ptr.port() {
+                prefix_url_str += ":";
+                prefix_url_str += &port.to_string();
+            }
+            prefix_url_str += "/";
+            prefix_url_str += params.get(PARAMETER_DATASET_NAME).unwrap();
+            prefix_url_str += "/";
+
+            let prefix_url = Url::parse(prefix_url_str.as_str()).unwrap();
+
+            ws_smart_transfer_protocol_axum_server::dataset_pull_ws_handler(socket, dataset, prefix_url)
+        }
+    )
+}
+
+async fn resolve_dataset(
+    local_dataset_repository: axum::extract::Extension<Arc<dyn LocalDatasetRepository>>,
+    params: &HashMap<String, String>
+) -> Result<Arc<dyn Dataset>, GetDatasetError> {
+    
+    // TODO: support 'accountName' parameter
+    let dataset_name_param = params.get(PARAMETER_DATASET_NAME).unwrap();
```

Comment:
What do you mean by "the routes don't mention the datasetName"? We have a nested router.
```
            .nest(
                format!("/:{}", kamu_adapter_http::PARAMETER_DATASET_NAME).as_str(),
                kamu_adapter_http::create_dataset_transfer_protocol_routes(local_repo),
            )
```

Reply:
If I understood the idea correctly, you suggest to split out the lower-level part of the route handlers from the operations implementation logic completely. So that current handler, i.e., `dataset_blocks_handler`, receives a dataset reference as an argument, constructed block hash as an argument, and does not even know of Axum existence.

With regards to the account name, that would be another row in the router, likely preceeding the current, and I assumed it would reuse the nested router, so it is isolated in the `create_dataset_transfer_protocol_routes`.

What you've described about "externalizing" sounds a lot like middleware layer that propagates custom state to the final handlers. Does [this](https://docs.rs/axum/latest/axum/middleware/index.html#passing-state-from-middleware-to-handlers) look similar?

Reply:
Yes, to be specific my suggestion is to write `kamu-adapter-http` in such a way that its routes can be mounted under any URL structure like:
- `/:datasetName/...` - single-tenant, by name
- `/:datasetID/...` - single-tenant, by ID
- `/:accountName/:datasetName/...` - multi-tenant, by name
- `/...` - or even a server that serves only one specific dataset

Therefore the logic that determines the identity of a dataset that STP handlers will operate on has to be defined at the place of routes **nesting**.

I'm not familiar enough with Axum's services/layers/middleware structure to say for sure, but what you found looks very close how I imagined it.

Reply:
I assume this is solved by your middleware branch now